File: c++_solutions/62.cpp
// simple DP problem
class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>>DP(m , vector<int>(n , 0)) ; 
        DP[0][0] = 1 ; 
        for(int i = 0 ; i < m ; i++){
            for(int j = 0 ; j < n ; j++){
                if(i !=0 || j!=0){
                    if(i-1>=0){
                        DP[i][j] += DP[i-1][j] ; 
                    }
                    if(j-1>=0){
                        DP[i][j] += DP[i][j-1] ; 
                    }
                }
            }
        }
        return DP[m-1][n-1] ; 
    }
};

File: python_solutions/62.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # Solution Overview
        # use dp to store the results and then update the next states
        cnt_path = [0 for col in range(n)]
        prev_path = cnt_path.copy()
        for row in range(m):
            for col in range(n):
                prev_path[col] = cnt_path[col]
                cnt_path[col] = 0
            for col in range(n):
                if row == 0 and col == 0:
                    cnt_path[col] = 1
                else:
                    if row-1 >= 0 :
                        cnt_path[col] += prev_path[col]
                    if col-1 >= 0 :
                        cnt_path[col] += cnt_path[col-1]
        return cnt_path[n-1]
